oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollectorIT.java

```diff
@@ -1182,6 +1201,362 @@ public void testDeletedPropsAndUnmergedBCWithCollision() throws Exception {
+        createNodes("/a/b/c/d");
+        lateWriteRemoveNodes(Arrays.asList("/a/b"), null);
+
+        assertTrue(getChildeNodeState(store1, "/a/b/c/d", true).exists());
+
+        // wait two hours
+        clock.waitUntil(clock.getTime() + HOURS.toMillis(2));
+        // clean everything older than one hour
+        VersionGCStats stats = gc(store1.getVersionGarbageCollector(), 1, HOURS);
+        assertNotNull(stats);
+
+        assertTrue(store1.getDocumentStore().find(NODES, "2:/a/b") != null);
+        assertTrue(store1.getDocumentStore().find(NODES, "4:/a/b/c/d") != null);
+        assertTrue(getChildeNodeState(store1, "/a/b/c/d", true).exists());
+        //TODO: below assert fails currently as uncommitted revisions aren't yet removed
+        // should be 3 as it should clean up the _deleted from /a/b, /a/b/c and /a/b/c/d
+        assertEquals(3, stats.updatedDetailedGCDocsCount);
+    }
+
+    /**
+     * This (re)produces a case where classic GC deletes nodes
+     * but they are still in the nodes cache, eg:
+     *  org.apache.jackrabbit.oak.plugins.document.ConflictException: 
+     * The node 4:/a/b/c/d does not exist or is already deleted 
+     * at base revision r2-0-1,r1-0-2,
+     * branch: null, commit revision: re-0-1]
+     */
+    @Test
+    @Ignore(value = "OAK-10658 : fails currently as invalidation is missing (in classic GC) after late-write-then-sweep-then-GC")
+    public void lateWriteRemoveChildGC_withSweep() throws Exception {
+        assumeTrue(fixture.hasSinglePersistence());
+        enableDetailedGC(store1);
+        createNodes("/a/b/c/d");
+        lateWriteRemoveNodes(Arrays.asList("/a/b"), "/foo");
+
+        getChildeNodeState(store1, "/a/b/c/d", true);
+
+        // wait two hours
+        clock.waitUntil(clock.getTime() + HOURS.toMillis(2));
+        // clean everything older than one hour
+        getChildeNodeState(store1, "/a/b/c/d", true);
+        assertTrue(store1.getDocumentStore().find(NODES, "4:/a/b/c/d") != null);
+        assertTrue(store1.getDocumentStore().find(NODES, "3:/a/b/c") != null);
+        assertTrue(store1.getDocumentStore().find(NODES, "2:/a/b") != null);
+
+        VersionGCStats stats = gc(store1.getVersionGarbageCollector(), 1, HOURS);
+        assertNotNull(stats);
+
+        assertFalse(store1.getDocumentStore().find(NODES, "4:/a/b/c/d") != null);
+        assertFalse(store1.getDocumentStore().find(NODES, "3:/a/b/c") != null);
+        assertFalse(store1.getDocumentStore().find(NODES, "2:/a/b") != null);
+
+        // invalidating store1's nodeCache would fix it
+        // but we need that to happen in prod code, not test
+//        store1.getNodeCache().invalidateAll();
+
+        // creating /a/b/c again, below late-write-removed /a/b
+        // triggered a ConflictException
+        createNodes("/a/b/c/d/e");
+        getChildeNodeState(store1, "/a/b/c/d/e", true);
+    }
+
+    private void enableDetailedGC(DocumentNodeStore dns) throws IllegalAccessException {
```

Comment:
This method could be moved to `DetailGCHelper` class.

Reply:
good point, done in https://github.com/apache/jackrabbit-oak/pull/1315/commits/0cca811570536841ac531d4d4ad7c00d2d20bce6